### ISSUE TITLE
[deploy] Check if the deployed artifacts are not in Maven Central already

### ISSUE THREAD

**Issue by gastaldi** [jreleaser/jreleaser#1721]
Trying to release the same artifact twice in Nexus gives a cryptic error, which is only visible in the logs. It would be nice if JReleaser could validate if the artifact(s) to be deployed are not already in central BEFORE attempting to deploy it and fail the build with a user-friendly message.
```
[INFO]      [nexus2] Creating a staging repository for io.quarkiverse
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3-javadoc.jar
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3-javadoc.jar.asc
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3-javadoc.jar.md5
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3-javadoc.jar.sha1
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3-javadoc.jar.sha256
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3-javadoc.jar.sha512
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3-sources.jar
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3-sources.jar.asc
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3-sources.jar.md5
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3-sources.jar.sha1
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3-sources.jar.sha256
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3-sources.jar.sha512
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3.jar
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3.jar.asc
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3.jar.md5
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3.jar.sha1
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3.jar.sha256
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3.jar.sha512
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3.pom
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3.pom.asc
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3.pom.md5
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3.pom.sha1
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3.pom.sha256
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-deployment/0.0.3/quarkus-slack-deployment-0.0.3.pom.sha512
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-parent/0.0.3/quarkus-slack-parent-0.0.3.pom
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-parent/0.0.3/quarkus-slack-parent-0.0.3.pom.asc
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-parent/0.0.3/quarkus-slack-parent-0.0.3.pom.md5
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-parent/0.0.3/quarkus-slack-parent-0.0.3.pom.sha1
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-parent/0.0.3/quarkus-slack-parent-0.0.3.pom.sha256
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack-parent/0.0.3/quarkus-slack-parent-0.0.3.pom.sha512
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3-javadoc.jar
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3-javadoc.jar.asc
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3-javadoc.jar.md5
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3-javadoc.jar.sha1
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3-javadoc.jar.sha256
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3-javadoc.jar.sha512
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3-sources.jar
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3-sources.jar.asc
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3-sources.jar.md5
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3-sources.jar.sha1
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3-sources.jar.sha256
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3-sources.jar.sha512
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3.jar
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3.jar.asc
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3.jar.md5
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3.jar.sha1
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3.jar.sha256
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3.jar.sha512
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3.pom
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3.pom.asc
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3.pom.md5
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3.pom.sha1
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3.pom.sha256
[INFO]      [nexus2]  - io/quarkiverse/slack/quarkus-slack/0.0.3/quarkus-slack-0.0.3.pom.sha512
[INFO]      [nexus2] Closing staging repository ioquarkiverse-3[136](https://github.com/quarkiverse-release/quarkiverse-release/actions/runs/10722253833/job/29732745129#step:9:142)
[INFO]      [nexus2] This operation may take some time. Please wait.
[INFO]      [nexus2] Attempt 1 of 61
[INFO]      [nexus2] Attempt 2 of 61
[INFO]      [nexus2] Attempt 3 of 61
[INFO]      [nexus2] Attempt 4 of 61
[INFO]      [nexus2] Releasing staging repository ioquarkiverse-3136
[INFO]      [nexus2] This operation may take some time. Please wait.
[INFO]      [nexus2] Attempt 1 of 61
[INFO]      [nexus2] Attempt 2 of 61
[INFO]  Writing output properties to out/jreleaser/output.properties
Error:  JReleaser failed after 01:24 m
```

**Comment by aalmiray** [comment]
This one may be a bit tricky. In theory it would be enough to issue a HEAD on `https://repo1.maven.org/maven2/${groupId}/${artifactId}/${version}/${filename}`, right? However, the `nexus2` deployer may be used to publish artifacts to _any_ Nexus2 instance, not just the one that is currently used to stage & publish to Maven Central.

This means adding a generic boolean property for checking if the artifact exists in Maven Central may not be a good idea. Also, publishing to MC via Nexus2 may use 2 different input URLs (old and new OSSRH) yet the HEAD call should be performed on a 3rd URL.

Perhaps adding a new property that takes an URL as input (for issuing the HEAD op) is the way to go. If a value is defined then it means the check should be performed before each artifact is deployed. If no value is defined then no check is performed.

Checks should be executed for all artifacts as there might have been a successful partial deployment. Issue a warning for any previously existing artifact.

Now, what should happen when at least 1 artifact has already been deployed in a previous session?
 - [ ] fail deployment after all artifacts haven been checked/uploaded
 - [ ] deployment successful if at least 1 artifact was deployed

If all artifacts were fully deployed then:
- [ ] fail deployment
- [ ] deployment skipped with a warning

@gastaldi LMKWYT

**Comment by aalmiray** [comment]
> A partial deployment is not something you'd see frequently, as a staging repository in Nexus is only released after being closed, and that happens when the whole deployment is done.
> 
> 
> 
> IMHO if an artifact has already been deployed (regardless if it happened in a previous session or deployed from somewhere else) the whole deployment should fail. That's what Nexus does remotely during the rules validation when you push to a staging repository. 

Understood. Any existing artifact will fail the whole deployment. However, I think it makes sense to query all potential deployables and list all existing as part of the failure message.

> I wonder if this validation could be part of the `pomChecker check-maven-central` command? 

No, it should not as pomchecker only checks the contents of a pom.xml. Single responsibility principle for this particular tool. 

JReleaser OTOH can and should perform this check. What do you think about the suggested uses of the URL property?

**Comment by gastaldi** [comment]
> JReleaser OTOH can and should perform this check. What do you think about the suggested uses of the URL property?

+1, sounds like a good addition.

**Comment by aalmiray** [comment]
🎉 This issue has been resolved in `v1.15.0` ([Release Notes](https://github.com/jreleaser/jreleaser/releases/tag/v1.15.0))
